test-harness/src/test/resources/nested_fork_join_swf.json

```diff
@@ -0,0 +1,104 @@
+{
+  "name": "nested_fork_join_swf",
+  "description": "nested_fork_join_swf",
+  "version": 1,
+  "tasks": [
+    {
+      "name": "outer_fork",
+      "taskReferenceName": "outer_fork",
+      "inputParameters": {
+        "param1": "${workflow.input.param1}",
+        "param2": "${workflow.input.param2}",
+        "subwf": "${workflow.input.nextSubwf}"
+      },
+      "type": "FORK_JOIN",
+      "forkTasks": [
+        [
+          {
+            "name": "inner_fork",
+            "taskReferenceName": "inner_fork",
+            "inputParameters": {
+              "param1": "${workflow.input.param1}",
+              "param2": "${workflow.input.param2}",
+              "subwf": "${workflow.input.nextSubwf}"
+            },
+            "type": "FORK_JOIN",
+            "forkTasks": [
+              [
+                {
+                  "name": "sub_workflow_task",
+                  "taskReferenceName": "st1",
+                  "inputParameters": {
+                    "param1": "${workflow.input.param1}",
+                    "param2": "${workflow.input.param2}",
+                    "subwf": "${workflow.input.nextSubwf}"
+                  },
+                  "type": "SUB_WORKFLOW",
+                  "subWorkflowParam": {
+                    "name": "${workflow.input.subwf}",
+                    "version": 1
+                  },
+                  "retryCount": 0
+                }
+              ],
+              [
+                {
+                  "name": "integration_task_2",
+                  "taskReferenceName": "t2",
+                  "inputParameters": {
+                    "p1": "${workflow.input.param1}",
+                    "p2": "${workflow.input.param2}"
+                  },
+                  "type": "SIMPLE",
+                  "retryCount": 0
+                }
+              ]
+            ]
+          },
+          {
+            "name": "inner_join",
+            "taskReferenceName": "inner_join",
+            "type": "JOIN",
+            "joinOn": [
+              "st1",
+              "t2"
+            ]
+          }
+        ],
+        [
+          {
+            "name": "integration_task_2",
+            "taskReferenceName": "t3",
+            "inputParameters": {
+              "p1": "${workflow.input.param1}",
+              "p2": "${workflow.input.param2}"
+            },
+            "type": "SIMPLE",
+            "retryCount": 0
+          }
+        ]
+      ]
+    },
+    {
+      "name": "join",
+      "taskReferenceName": "outer_join",
+      "inputParameters": {},
+      "type": "JOIN",
+      "joinOn": [
+        "outer_fork",
```

Comment:
this should be "inner_join"

Reply:
Is this an unwritten rule or convention? Using `outer_fork` in the `JOIN` task works as expected in that the `outer_join` completes before `inner_join` is completed.

Reply:
Not an unwritten rule but 

>Using outer_fork in the JOIN task works as expected in that the outer_join completes before inner_join is completed.

It completes in the expected way in the test setup because t3 completes alongwith t2 and the state is checked after. However, if t3 is completed while t2 is still IN_PROGRESS, the outer_join would now be COMPLETED because its dependent tasks `outer_fork` and `t3` are in COMPLETED state.